### PR TITLE
Fixes #33909 - Drop data migrations before Foreman 2.0 

### PR DESCRIPTION
**Pull request by tbrisker**
All data migrations only modify data on upgrades. On new installs they
are not required. Since we do not generally support upgrading multiple
releases, we can clean up old data migrations. This commit cleans up all
the data migrations that were present before Foreman 2.0.

Second commit, since I know this topic may be controversial:

Refs #33909 - Delete empty migrations

New installs will not run these empty migrations, and old installs will
only list that they have been executed when checking the `rake
db:migrate:status` with no file, which is harmless.

### CODE REVIEW

**Comment by ezr-ondrej** [comment]
Could we add some purge task to remove the migrations from schema_migrations? It would be my only concern with deleting migrations - having a way to remove them from DB (and also having some list of such)

File: db/seeds.d/110-smart_proxy_features.rb [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Proxy features
-proxy_features = ["TFTP", "DNS", "DHCP", "Puppet CA", "BMC", "Realm", "Facts", "Logs", "HTTPBoot", "External IPAM",
+proxy_features = ["Templates", "TFTP", "DNS", "DHCP", "Puppet CA", "BMC", "Realm", "Facts", "Logs", "HTTPBoot", "External IPAM",
```
Comment (ezr-ondrej) [inline db/seeds.d/110-smart_proxy_features.rb:2]: Is this related?

**Comment by tbrisker** [comment]
@ezr-ondrej we could, but I don't see much benefit in doing so - if we keep them, we can tell that they ran in the past even if they no longer exist (i.e. identify if the db is new or was upgraded). 
Having timestamps in schema_migrations without files is valid. Many rails apps delete all old migrations, when the `schema.rb` or `structure.sql` file is committed in source control and is used as the base db structure - which unfortunately we cannot do due to $reasons.

File: db/seeds.d/110-smart_proxy_features.rb (hunk 1 above)
Comment (tbrisker) [reply db/seeds.d/110-smart_proxy_features.rb:2]: yes, It was added in a migration previously instead of the seed.

File: db/migrate/20100310080727_add_family_to_os.rb [hunk 2]
```diff
@@ -1,29 +1,6 @@
...
-
   def up
     add_column :operatingsystems, :family_id, :integer
-
-    Operatingsystem.reset_column_information
-
-    Operatingsystem.all.each do |os|
-      case os.name
-      when /RedHat|Centos|Fedora/i
-        os.family_id = Operatingsystem::FAMILIES.index :RedHat
```
Comment (ekohl) [inline db/migrate/20100310080727_add_family_to_os.rb:12]: Fun fact: I can't see how this worked. AFAIK Families never had an entry which accepted symbols and it always was `Redhat`, not `RedHat`.

File: db/migrate/20121118120028_migrate_hypervisors_to_compute_resources.rb [hunk 3]
```diff
@@ -1,18 +1,6 @@
...
-
-      Foreman::Model::Libvirt.create :name        => hypervisor.name,
-                                     :url         => hypervisor.uri,
-                                     :description => "Automatically migrated from hypervisor #{hypervisor.name} / #{hypervisor.uri}"
-    end
-    drop_table :hypervisors
+    drop_table :hypervisors, if_exists: true
   end
 
   def down
```
Comment (ekohl) [inline db/migrate/20121118120028_migrate_hypervisors_to_compute_resources.rb:6]: It's weird to see an `up` migration that's a `drop_table` but an empty `down`.

File: db/migrate/20140902102858_convert_ipam_to_string.rb [hunk 4]
```diff
@@ -1,30 +1,14 @@
 class ConvertIpamToString < ActiveRecord::Migration[4.2]
-  class FakeSubnet < ApplicationRecord
-    self.table_name = 'subnets'
-  end
-
   def up
     add_column :subnets, :ipam_tmp, :string, :default => IPAM::MODES[:dhcp], :null => false, :limit => 255
```
Comment (ekohl) [inline db/migrate/20140902102858_convert_ipam_to_string.rb]: If you're dropping the migration, there's no reason for a tmp column. Can't this be squashed down to a `change_column`? Same comment for `down`.

File: db/migrate/20190226182746_create_registration_facets.rb [hunk 5]
```diff
@@ -1,5 +1,3 @@
-# frozen_string_literal: true
-
```
Comment (ekohl) [inline db/migrate/20190226182746_create_registration_facets.rb:2]: I'm not sure if this is needed.

File: db/migrate/20210114143800_nilify_empty_proxy_credentials.rb [hunk 6]
```diff
@@ -1,19 +1,11 @@
 class NilifyEmptyProxyCredentials < ActiveRecord::Migration[6.0]
   def up
-    if User.unscoped.find_by_login(User::ANONYMOUS_ADMIN).present?
-      User.as_anonymous_admin do
```
Comment (ekohl) [inline db/migrate/20210114143800_nilify_empty_proxy_credentials.rb:4]: Was this here to disable auditing or something? This is fairly recent (d60d1cb169ac83a8aed6ba7256c1b731dce07350) so maybe @ares remembers.

**Review by ekohl — COMMENTED** [review]
I still would have preferred to really squash the migrations so avoid changing tables, but I can live with this as a start.

File: db/migrate/20121118120028_migrate_hypervisors_to_compute_resources.rb (hunk 3 above)
Comment (tbrisker) [reply db/migrate/20121118120028_migrate_hypervisors_to_compute_resources.rb:6]: That's true, although it was that way before as well so no idea what should be in that table. Also, I highly doubt anyone would rollback migrations all they way to 2012 😜

File: db/migrate/20190226182746_create_registration_facets.rb (hunk 5 above)
Comment (tbrisker) [reply db/migrate/20190226182746_create_registration_facets.rb:2]: I don't know why it was added in the first place, but there are no strings in this file so it had no affect anyways.

File: db/migrate/20210114143800_nilify_empty_proxy_credentials.rb (hunk 6 above)
Comment (tbrisker) [reply db/migrate/20210114143800_nilify_empty_proxy_credentials.rb:4]: I'm not sure, possibly an attempt to bypass taxonomy or authorization scope. `update_all` triggers sql `UPDATE` query directly without initializing AR objects so it wouldn't have been audited in any case.

**Comment by ekohl** [comment]
Tests fail with:
```
NoMethodError: undefined method `find_by_lower_login' for #<Class:0x0000000009c55fb8>
Did you mean?  find_by_login
```

**Comment by tbrisker** [comment]
I've re-added the column info reset to the migration that adds lower_login to users. While the specific migration that fails should be fixed by GH-8918, there could potentially be other migrations in plugins that try looking up users (which they shouldn't, but that's outside the scope of this PR to fix).

**Review by ezr-ondrej — APPROVED** [review]
I agree that ultimate goal should be migration squash, but this is definitelly a step in the right direction. So big :+1:

**Review by ekohl — APPROVED** [review]
I wonder if this has a (positive) performance impact or if it's negligible. I'll see this as a step in the right direction to actually squashing.

**Comment by ezr-ondrej** [comment]
Thanks @tbrisker ! :)
